service/storage_service.cc

```diff
@@ -4412,7 +4412,62 @@ future<> storage_service::move(token new_token) {
 
 future<std::vector<storage_service::token_range_endpoints>>
 storage_service::describe_ring(const sstring& keyspace, bool include_only_local_dc) const {
-    return locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+    size_t nr_tablet_tables = 0;
+    co_await _db.local().get_tables_metadata().for_each_table_gently(coroutine::lambda([&] (table_id id, lw_shared_ptr<replica::table> table) -> future<> {
+        auto s = table->schema();
+        if (s->ks_name() == keyspace && table->uses_tablets()) {
+            ++nr_tablet_tables;
+        }
+        co_return;
+    }));
+    if (nr_tablet_tables) {
+        throw std::runtime_error(format("The keyspace {} has tablet table. Query describe_ring with the table parameter!", keyspace));
+    }
+    co_return co_await locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+}
+
+future<std::vector<dht::token_range_endpoints>>
+storage_service::describe_ring_for_table(const sstring& keyspace_name, const sstring& table_name) const {
+    slogger.debug("describe_ring for table {}.{}", keyspace_name, table_name);
+    auto& t = _db.local().find_column_family(keyspace_name, table_name);
+    if (!t.uses_tablets()) {
+        auto ranges = co_await describe_ring(keyspace_name);
+        co_return ranges;
+    }
```

Comment:
so, if the table in question does not use tablets, we just describe the whole keyspace, and drop the "table_name" on floor? i am wondering what's the merit of stuffing these two APIs into a single one for supporting both the tablets and vnode replication strategies? i think, scylla manager would want to use different workflows to perform cleanup on these two types of keyspaces anyway. and IMHO, the RESTful API would be more structured if we can expose API with the resource being accessed in a structured way, like "/tables/<keyspace>/<table>/token_ranges" .

Reply:
Yes, vnode tables in the same keyspace will have the same describe ring result.  For the table parameter,  we could use path parameter, but in most of the other options, we use query parameter.  Especially, if there are lots of parameters, it will be hard to use path parameters. 

Yes, we might have different workflows for the management stack to work with vnode and tablet tables, but it does not mean we should introduce more differences where it is not necessary.

Reply:
fair enough, i was thinking from the perspective of RESTful API design.